bluefog/common/mpi_controller.cc

```diff
@@ -571,20 +669,41 @@ void MPIController::NeighborAllreduce(std::vector<TensorTableEntry>& entries) {
 
   if (!first_entry.is_hierarchical) {
     if (!first_entry.dynamic_neighbors_enabled) {
-      int ret_code = MPI_Neighbor_allgather(
+      MPICHECK(MPI_Neighbor_allgather(
           fused_input_data, num_elements, mpi_ctx_.GetMPIDataType(first_entry.tensor),
           buffer_data, num_elements, mpi_ctx_.GetMPIDataType(first_entry.output),
-          mpi_ctx_.GetMPICommunicator(Communicator::GRAPH));
-      if (ret_code != MPI_SUCCESS) {
-        throw std::runtime_error(
-            "MPI_Neighbor_allreduce (through neighbor_allgather) failed, see MPI "
-            "output for details.");
-      }
+          mpi_ctx_.GetMPICommunicator(Communicator::GRAPH)));
     } else {
-      error_message = mpi_ctx_.NeighborValueExchangeWithConstantElements(
-        fused_input_data, buffer_data, num_elements, first_entry.output->dtype(),
-        first_entry.send_neighbors.get(), first_entry.recv_neighbors.get()
-      );
+      int nsend = first_entry.send_neighbors->size();
+      int nrecv = first_entry.recv_neighbors->size();
+      std::vector<MPI_Request> requests(nsend + nrecv);
+      std::vector<MPI_Status> statuses(nsend + nrecv);
+      for (int i = 0; i < nrecv; ++i) {
+        void* recvbuf =
+            (void*)((uint8_t*)buffer_data + num_elements * i * element_size);
+        MPICHECK(MPI_Irecv(recvbuf, num_elements,
+            mpi_ctx_.GetMPIDataType(first_entry.output), first_entry.recv_neighbors->at(i),
+            /*tag=*/mpi_ctx_.rank_ + first_entry.recv_neighbors->at(i),
+            mpi_ctx_.GetMPICommunicator(Communicator::GLOBAL), &requests[i + nsend]));
+      }
+#if HAVE_CUDA
+      if (first_entry.dst_weighting_enabled && first_entry.device != CPU_DEVICE_ID) {
+        cudaStreamSynchronize(mpi_ctx_.stream);
+      }
+#endif
+      for (int i = 0; i < nsend; ++i) {
+        const void* sendbuf = fused_input_data;
+        if (first_entry.dst_weighting_enabled)
+          sendbuf = 
+              (void*)((uint8_t*)weighted_fused_input_data + num_elements * i * element_size);
+        MPICHECK(MPI_Isend(sendbuf, num_elements,
+            mpi_ctx_.GetMPIDataType(first_entry.tensor), first_entry.send_neighbors->at(i),
+            /*tag=*/mpi_ctx_.rank_ + first_entry.send_neighbors->at(i),
+            mpi_ctx_.GetMPICommunicator(Communicator::GLOBAL), &requests[i]));
+      }
+      MPI_Waitall(nsend + nrecv, requests.data(), statuses.data());
+      error_message =
+          GenerateNeighborExchangeErrorMessage(statuses, nsend, nrecv);
     }
   } else {
```

Comment:
I think we should start to add some comments in long `if{} else{}` place that `else` is else to what condition.

Reply:
Done.